ambry-router/src/main/java/com.github.ambry.router/GetBlobInfoOperation.java

```diff
@@ -0,0 +1,243 @@
+      } else {
+        // the entries are ordered by correlation id and time. Break on the first request that has not timed out.
+        break;
+      }
+    }
+
+    checkAndMaybeComplete();
+    if (isOperationComplete()) {
+      return;
+    }
+
+    Iterator<ReplicaId> replicaIterator = operationTracker.getReplicaIterator();
+    while (replicaIterator.hasNext()) {
+      ReplicaId replicaId = replicaIterator.next();
+      String hostname = replicaId.getDataNodeId().getHostname();
+      Port port = new Port(replicaId.getDataNodeId().getPort(), PortType.PLAINTEXT);
+      GetRequest getRequest = createGetRequest(blobId, getOperationFlag());
+      RequestInfo request = new RequestInfo(hostname, port, getRequest);
+      int correlationId = getRequest.getCorrelationId();
+      correlationIdToGetRequestInfo.put(correlationId, new GetRequestInfo(replicaId, time.milliseconds()));
+      requestRegistrationCallback.registerRequestToSend(this, request);
+      replicaIterator.remove();
+    }
+  }
+
+  /**
+   * Handle the body of the response: Deserialize and set the {@link BlobInfo} to return.
+   * @param payload the body of the response.
+   * @throws IOException if there is an IOException while deserializing the body.
+   * @throws MessageFormatException if there is a MessageFormatException while deserializing the body.
+   */
+  void handleBody(InputStream payload)
+      throws IOException, MessageFormatException {
+    if (blobInfo == null) {
+      blobInfo = new BlobInfo(MessageFormatRecord.deserializeBlobProperties(payload),
+          MessageFormatRecord.deserializeUserMetadata(payload).array());
+    } else {
+      // If the successTarget is 1, this case will never get executed.
+      // If it is more than 1, then, different responses will have to be reconciled in some way. Here is where that
+      // would be done. Since the store is immutable, currently we handle this by ignoring subsequent responses
+      // (if the successTarget is > 1).
+    }
+  }
+
+  /**
+   * Handle the given {@link ResponseInfo} and update the status of the operation.
+   * Based on the status of the response, this involves handling the body, notifying the operation tracker,
+   * notifying the response handler and possibly setting the operationException.
+   * @param responseInfo the {@link ResponseInfo} to be handled.
+   */
+  @Override
+  void handleResponse(ResponseInfo responseInfo) {
+    if (isOperationComplete()) {
+      return;
+    }
+    int correlationId = ((GetRequest) responseInfo.getRequest()).getCorrelationId();
+    // Get the GetOperation that generated the request.
+    GetRequestInfo getRequestInfo = correlationIdToGetRequestInfo.remove(correlationId);
+    if (getRequestInfo == null) {
+      // Ignore right away. This associated operation has completed.
+      return;
+    }
+    if (responseInfo.getError() != null) {
+      setOperationException(new RouterException("Operation timed out", RouterErrorCode.OperationTimedOut));
```

Comment:
The network client can only really error out if it could not check out a connection for sending the request, at least as of this time, and therefore I think `OperationTimedOut` is a better error.